apps/prairielearn/assets/scripts/instructorFileEditorClient.tsx

```diff
@@ -4,12 +4,81 @@ import prettierEstreePlugin from 'prettier/plugins/estree';
 import * as prettier from 'prettier/standalone';
 
 import { onDocumentReady } from '@prairielearn/browser-utils';
+import { renderHtml } from '@prairielearn/preact';
 
 import { b64DecodeUnicode, b64EncodeUnicode } from '../../src/lib/base64-util.js';
+import { type FileMetadata, FileType } from '../../src/lib/editorUtil.types.js';
 
 import { configureAceBasePaths } from './lib/ace.js';
 import './lib/verboseToggle.js';
 
+/**
+ * Error codes for save validation issues.
+ */
+enum SaveErrorCode {
+  INVALID_JSON = 'INVALID_JSON',
+  UUID_CHANGED = 'UUID_CHANGED',
+  UUID_REMOVED = 'UUID_REMOVED',
+}
+
+/**
+ * Modal content component for invalid JSON error.
+ */
+function InvalidJsonModalContent() {
+  return (
+    <div class="alert alert-danger d-flex flex-column align-items-start mb-0">
+      <div class="d-flex flex-row align-items-start gap-2 mb-1">
+        <i class="bi bi-x-circle-fill fs-6" />
+        <strong>Invalid JSON</strong>
+      </div>
+      <div>
+        This file contains invalid JSON syntax and cannot be saved. Please fix the errors before
+        saving.
+      </div>
+    </div>
+  );
+}
+
+/**
+ * Modal content component for UUID change warning.
+ */
+function UuidChangeModalContent({
+  errorCode,
+  originalUuid,
+  newUuid,
+}: {
+  errorCode: SaveErrorCode.UUID_CHANGED | SaveErrorCode.UUID_REMOVED;
+  originalUuid?: string;
+  newUuid?: string;
+}) {
+  const getMessage = () => {
+    if (errorCode === SaveErrorCode.UUID_CHANGED && originalUuid && newUuid) {
+      return (
+        <>
+          The UUID in this file was changed from <code>"{originalUuid}"</code> to{' '}
+          <code>"{newUuid}"</code>.
+        </>
+      );
+    } else if (errorCode === SaveErrorCode.UUID_REMOVED) {
+      return <>The UUID was removed from this file.</>;
+    }
+    return <>The UUID was modified.</>;
+  };
+
+  return (
+    <>
+      <div class="alert alert-warning d-flex flex-column mb-3">
+        <div class="d-flex flex-row align-items-start gap-2 mb-1">
+          <i class="bi bi-exclamation-triangle-fill fs-6" />
+          <strong>UUID Change</strong>
+        </div>
+        <div>{getMessage()}</div>
+      </div>
+      <div class="ms-0">Hitting "Confirm Save" will save this file with its original UUID.</div>
```

Comment:
A `div` doesn't have any default margins that need to be removed:

```suggestion
      <div>Hitting "Confirm Save" will save this file with its original UUID.</div>
```